segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -521,66 +577,68 @@ private void persistMetadata(LogMetadata metadata, boolean create) throws Durabl
      * 5. Close the previous ledger (and implicitly seal it).
      * 5.1 If closing fails, there is nothing we can do. We've already opened a new ledger and new writes are going to it.
+     *
+     * NOTE: this method is not thread safe and is not meant to be executed concurrently. It should only be invoked as
+     * part of the Rollover Processor.
      */
-    private void rollover() throws DurableDataLogException {
+    private void rollover() {
         long traceId = LoggerHelpers.traceEnterWithContext(log, this.traceObjectId, "rollover");
+        val l = getWriteLedger().ledger;
+        if (l.getLength() < this.config.getBkLedgerMaxSize()) {
+            LoggerHelpers.traceLeave(log, this.traceObjectId, "rollover", traceId, false);
+            return;
+        }
 
-        // Create new ledger.
-        LedgerHandle newLedger = Ledgers.create(this.bookKeeper, this.config);
-        log.debug("{}: Rollover: created new ledger {}.", this.traceObjectId, newLedger.getId());
+        try {
+            // Create new ledger.
+            LedgerHandle newLedger = Ledgers.create(this.bookKeeper, this.config);
+            log.debug("{}: Rollover: created new ledger {}.", this.traceObjectId, newLedger.getId());
 
-        // Update the metadata.
-        LogMetadata metadata = getLogMetadata();
-        metadata = updateMetadata(metadata, newLedger);
-        LedgerMetadata ledgerMetadata = metadata.getLedger(newLedger.getId());
-        assert ledgerMetadata != null : "cannot find newly added ledger metadata";
-        log.debug("{}: Rollover: updated metadata '{}.", this.traceObjectId, metadata);
+            // Update the metadata.
+            LogMetadata metadata = getLogMetadata();
+            metadata = updateMetadata(metadata, newLedger);
+            LedgerMetadata ledgerMetadata = metadata.getLedger(newLedger.getId());
+            assert ledgerMetadata != null : "cannot find newly added ledger metadata";
+            log.debug("{}: Rollover: updated metadata '{}.", this.traceObjectId, metadata);
 
-        // Update pointers to the new ledger and metadata.
-        LedgerHandle oldLedger;
-        synchronized (this.lock) {
-            oldLedger = this.writeLedger.ledger;
-            this.writeLedger = new WriteLedger(newLedger, ledgerMetadata);
-            this.logMetadata = metadata;
+            // Update pointers to the new ledger and metadata.
+            LedgerHandle oldLedger;
+            synchronized (this.lock) {
+                oldLedger = this.writeLedger.ledger;
+                this.writeLedger = new WriteLedger(newLedger, ledgerMetadata);
+                this.logMetadata = metadata;
+            }
+
+            // Close the old ledger. This must be done outside of the lock, otherwise the pending writes (and their callbacks
+            // will be invoked within the lock, thus likely candidates for deadlocks).
+            Ledgers.close(oldLedger);
```

Comment:
As discussed offline, we shall do this in a separate task. Issue #1414 was made for tracking.